gapis/api/vulkan/footprint_builder.go

```diff
@@ -0,0 +1,3283 @@
+		pipeline:       newLabel(),
+		dynamicState:   newLabel(),
+	}
+}
+
+type queueExecutionInfo struct {
+	currentCmdBufState   *commandBufferExecutionState
+	primaryCmdBufState   *commandBufferExecutionState
+	secondaryCmdBufState *commandBufferExecutionState
+
+	subpasses       []subpassInfo
+	subpass         *subpassIndex
+	renderPassBegin *forwardPairedLabel
+
+	currentCommand api.SubCmdIdx
+
+	framebuffer *FramebufferObject
+
+	lastSubmitID      api.CmdID
+	currentSubmitInfo *queueSubmitInfo
+}
+
+func newQueueExecutionInfo(id api.CmdID) *queueExecutionInfo {
+	return &queueExecutionInfo{
+		subpasses:      []subpassInfo{},
+		lastSubmitID:   id,
+		currentCommand: api.SubCmdIdx([]uint64{0, 0, 0, 0}),
+	}
+}
+
+func (qei *queueExecutionInfo) updateCurrentCommand(ctx context.Context,
+	fci api.SubCmdIdx) {
+	switch len(fci) {
+	case 4:
+		current := api.SubCmdIdx(qei.currentCommand[0:3])
+		comming := api.SubCmdIdx(fci[0:3])
+		if current.LessThan(comming) {
+			// primary command buffer changed
+			qei.primaryCmdBufState = newCommandBufferExecutionState()
+		}
+		qei.currentCmdBufState = qei.primaryCmdBufState
+	case 6:
+		if len(qei.currentCommand) != 6 {
+			// Transit from primary command buffer to secondary command buffer
+			qei.secondaryCmdBufState = newCommandBufferExecutionState()
+		} else {
+			current := api.SubCmdIdx(qei.currentCommand[0:5])
+			comming := api.SubCmdIdx(fci[0:5])
+			if current.LessThan(comming) {
+				// secondary command buffer changed
+				qei.secondaryCmdBufState = newCommandBufferExecutionState()
+			}
+		}
+		qei.currentCmdBufState = qei.secondaryCmdBufState
+	default:
+		log.E(ctx, "Invalid length of full command index")
+	}
+	qei.currentCommand = fci
+}
+
+func (qei *queueExecutionInfo) startSubpass(ctx context.Context,
+	bh *dependencygraph.Behaviour) {
+	write(ctx, bh, qei.subpass)
+	subpassI := qei.subpass.val
+	handleLoadOp := func(ctx context.Context, bh *dependencygraph.Behaviour,
+		attachment *subpassAttachmentInfo) {
+		// TODO: Not all subpasses change layouts
+		modify(ctx, bh, attachment.layout)
+		switch attachment.desc.LoadOp {
+		case VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_CLEAR,
+			VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_DONT_CARE:
+			if attachment.fullImageData {
+				write(ctx, bh, attachment.data)
+			} else {
+				modify(ctx, bh, attachment.data)
+			}
+		default:
+			read(ctx, bh, attachment.data)
+		}
+	}
+	for _, c := range qei.subpasses[subpassI].colorAttachments {
+		handleLoadOp(ctx, bh, c)
+	}
+	for _, r := range qei.subpasses[subpassI].resolveAttachments {
+		handleLoadOp(ctx, bh, r)
+	}
+	for _, i := range qei.subpasses[subpassI].inputAttachments {
+		handleLoadOp(ctx, bh, i)
+	}
+	if qei.subpasses[subpassI].depthStencilAttachment != nil {
+		dsAtt := qei.subpasses[subpassI].depthStencilAttachment
+		// TODO: Not all subpasses change layouts
+		modify(ctx, bh, dsAtt.layout)
+		if (dsAtt.desc.LoadOp == VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_CLEAR ||
```

Comment:
I'm wondering to which file I should add such helpers. Should I add them in the API file as subroutines (but cpp codes is not used), or in the externs.go (make more sense but still far away from the uses), or just in the footprint_builder.go (and make it just a private helper)?